### PR TITLE
Add TypedDict generics as previously opened in PR #446

### PR DESCRIPTION
**Pull request by tsmith023**
This PR adds generic capability to the client.collection API for users to specify custom classes that inherit from TypedDict in order to customise the expected returns of their method calls.

The primary decision was to isolate ORM-like behaviour into the client.collection_model API and leave the classic API as TypeScript-y as possible. As such, all data structures in `client.collection` are dictionaries with custom type sprinkling allowed through TypedDict at the end without the need for heavy serialization/deserialization of JSON to classes.

### CODE REVIEW

File: integration/test_collection.py [hunk 1]
```diff
@@ -61,18 +68,102 @@ def test_create_and_delete(client: weaviate.Client):
...
+        name: str
+
+    client.collection.get(name, Right)
+    client.collection.get(name, AlsoRight)
+    assert client.collection.exists(name)
     client.collection.delete(name)
+    assert not client.collection.exists(name)
+
+
+def test_get_with_unsupported_generic(client: weaviate.Client):
```
Comment (dirkkul) [inline integration/test_collection.py:93]: let's split this test up

File: integration/test_collection.py [hunk 2]
```diff
@@ -614,6 +705,104 @@ def test_mono_references_grcp(client: weaviate.Client):
...
+    C.data.insert(CPropsInsert(name="find me", age=10, ref=ReferenceTo(uuids=uuid_B)))
+
+    class BPropsGet(TypedDict):
+        name: str
+        ref: Reference[AProps]
+
+    class CPropsGet(TypedDict):
+        name: str
+        ref: Reference[BPropsGet]
+        not_specified: str
```
Comment (dirkkul) [inline integration/test_collection.py:772]: both of these are not used?

File: integration/test_collection.py [hunk 3]
```diff
@@ -614,6 +705,104 @@ def test_mono_references_grcp(client: weaviate.Client):
...
+
+    A = client.collection.create(
+        CollectionConfig(
+            name="ATypedDicts",
+            vectorizer=Vectorizer.NONE,
+            properties=[
+                Property(name="Name", data_type=DataType.TEXT),
+            ],
+        )
+    )
```
Comment (dirkkul) [inline integration/test_collection.py:724]: you didn't add the class here

File: weaviate/collection/classes/internal.py [hunk 4]
```diff
@@ -37,3 +42,12 @@ def _metadata_from_dict(metadata: Dict[str, Any]) -> _MetadataReturn:
         score=metadata.get("score"),
         is_consistent=metadata.get("isConsistent"),
     )
+
+
+def _extract_props_from_list_of_objects(type_: Any) -> Optional[Any]:
```
Comment (dirkkul) [inline weaviate/collection/classes/internal.py:47]: no usage?

File: weaviate/collection/classes/orm.py [hunk 5]
```diff
@@ -33,7 +33,7 @@
 
 @dataclass
 class CrossReference:
-    ref_type: Union[Type, str]
+    ref_type: Union[type, str]
```
Comment (dirkkul) [inline weaviate/collection/classes/orm.py:37]: why this change? Also 3 lines down it still uses "type"

File: weaviate/collection/collection.py [hunk 6]
```diff
@@ -1,67 +1,86 @@
...
 
 
-class CollectionObject:
+class CollectionObject(Generic[Properties]):
     def __init__(
         self,
         connection: Connection,
         name: str,
-        config: _ConfigCollection,
+        type_: Optional[Type[Properties]] = None,
```
Comment (dirkkul) [inline weaviate/collection/collection.py:20]: can we rename it to data_model everywhere for consistency?

**Review by dirkkul — COMMENTED** [review]
I like it!

But tested the code and I have a lot of Any in it:
Screenshot 2023-08-30 at 08 04 50

File: weaviate/collection/data.py [hunk 7]
```diff
@@ -302,22 +314,42 @@ def __parse_properties_grpc(data: Dict[str, Any]) -> weaviate_pb2.BatchObject.Pr
...
-    def _json_to_object(self, obj: Dict[str, Any]) -> _Object:
+class _DataCollection(Generic[Properties], _Data):
+    def __init__(
+        self,
+        connection: Connection,
+        name: str,
+        config: _ConfigBase,
+        consistency_level: Optional[ConsistencyLevel],
+        tenant: Optional[str],
+        type_: Optional[Type[Properties]] = None,
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collection/data.py:325]: This is an internal definition and I was thinking `type_` was clearer than `data_model`, my thought was that `data_model` is for the public API

File: integration/test_collection.py [hunk 8]
```diff
@@ -614,6 +705,104 @@ def test_mono_references_grcp(client: weaviate.Client):
...
+                    LinkTo(
+                        link_on="ref",
+                        properties=["name"],
+                        metadata=MetadataQuery(uuid=True),
+                    ),
+                ],
+                metadata=MetadataQuery(uuid=True, last_update_time_unix=True),
+            ),
+        ],
+        # type_=CPropsGet,
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply integration/test_collection.py:792]: Ah yes, this is because the gRPC refactoring isn't in this PR so I've left the generics in queries for later. We should also discuss this aspect as I don't remember a clear decision on the desired implementation

File: integration/test_collection.py [hunk 9]
```diff
@@ -61,18 +68,102 @@ def test_create_and_delete(client: weaviate.Client):
...
+    )
+
+    class YetAnotherWrong(BaseModel):
+        name: str
+
+    with pytest.raises(TypeError) as error:
+        client.collection.get("NotImportant", YetAnotherWrong)
+    assert (
+        error.value.args[0]
+        == "The only generics allowed to be used in data_model are Dicts and TypedDicts"
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply integration/test_collection.py:127]: The requirement is in fact harder in that it's "TypedDict or a class without an `__init__` method", i.e. something that can't be instantiated

File: weaviate/collection/classes/internal.py (hunk 4 above)
Comment (tsmith023) [reply weaviate/collection/classes/internal.py:47]: Relic of large PR, it will come back later but will delete for now :)

File: integration/test_collection.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dirkkul) [reply integration/test_collection.py:127]: that would also be clear :)

File: weaviate/collection/classes/orm.py (hunk 5 above)
Comment (tsmith023) [reply weaviate/collection/classes/orm.py:37]: Potentially a relic mypy change, will redo

**Comment by tsmith023** [comment]
I'm not sure why your IDE is not completing the type inference as mine does
Screenshot 2023-08-30 at 08 25 40

**Comment by dirkkul** [comment]
hmmm, I also used VSCode

**Comment by tsmith023** [comment]
Ah, it's because your function signature is incorrect. Your type Right is an arg following the kwarg config=collection_config, which is invalid syntax. You should do .create(collection_config, Right) or .create(config=collection_config, data_model=Right)

File: weaviate/collection/collection.py (hunk 6 above)
Comment (tsmith023) [reply weaviate/collection/collection.py:20]: I think `type_` makes more sense than `data_model` in the internal context with `data_model` just acting as the user-preferred nomenclature

Since Python calls its module `typing`, the type hint of `type_` is `Type[...]`, within `_DataCollection` the `type_` argument is used in a `get_type_hints` function call

I also imagine calling this `data_model` might become confusing when working with the ORM where it will either be just `model` or also `data_model`, which would be confusing in my book. Calling it `type_` here makes it clear that it is only a type hint and not an ORM model

File: integration/test_collection.py (hunk 2 above)
Comment (tsmith023) [reply integration/test_collection.py:772]: Is it okay if I comment this out for now and then add it back once the query generics syntax is agreed upon?

Comment (dirkkul) [reply integration/test_collection.py:772]: yes

File: integration/test_collection.py [hunk 10]
```diff
@@ -61,18 +68,116 @@ def test_create_and_delete(client: weaviate.Client):
...
+
+def test_get_with_pydantic_class(client: weaviate.Client):
+    class Wrong(BaseModel):
+        name: str
+
+    with pytest.raises(TypeError) as error:
+        client.collection.get("NotImportant", Wrong)
+    assert (
+        error.value.args[0]
+        == "data_model can only be a dict or a class that inherits from TypedDict"
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply integration/test_collection.py:139]: Ah okay, perhaps just `Optional[TypedDict]` since passing `None` defaults to `Dict[str, Any]` anyway. So it would be
>If provided, `data_model` can only be a class that inherits from `TypedDict`

The thing is, they could provide `Dict[str, str]` if they know their dictionary is only made of string values but don't want to use a `TypedDict`. Seems like an niche use-case but it should be supportable nonetheless

[the remark this answers is not in the document]
Comment (tsmith023) [reply integration/test_collection.py]: How's this?
>data_model can only be a dict type, e.g. Dict[str, str], or a class that inherits from TypedDict

[the remark this answers is not in the document]
Comment (dirkkul) [reply integration/test_collection.py]: perfect!

File: integration/test_collection.py [hunk 11]
```diff
@@ -68,8 +68,23 @@ def test_create_and_delete(client: weaviate.Client):
...
+        name=name,
+        properties=[Property(name="Name", data_type=DataType.TEXT)],
+        vectorizer=Vectorizer.NONE,
+    )
+    client.collection.create(collection_config)
+
+    client.collection.get(name, Dict[str, str])
+    assert client.collection.exists(name)
+    client.collection.delete(name)
+    assert not client.collection.exists(name)
```
Comment (dirkkul) [inline integration/test_collection.py:83]: this tests the create call, not the get call with the Dict[]

File: integration/test_collection.py [hunk 12]
```diff
@@ -86,34 +101,29 @@ class Right(TypedDict):
     assert not client.collection.exists(name)
 
 
-def test_get_with_empty_class(client: weaviate.Client):
+wrong_generic_error_msg = "data_model can only be a dict type, e.g. Dict[str, str], or a class that inherits from TypedDict"
```
Comment (dirkkul) [inline integration/test_collection.py:104]: constant are upper case

File: integration/test_collection.py (hunk 11 above)
Comment (tsmith023) [reply integration/test_collection.py:83]: Ah gotcha, `.create()` uses `.get()` under-the-hood but the test doesn't know that!

Comment (dirkkul) [reply integration/test_collection.py:83]: but still the difference is in the colelction that gets returned - so I would expect that the collection is tested - eg an insert + a return

Comment (dirkkul) [reply integration/test_collection.py:83]: > This is a good time to note that test_collection.py should be split up into collection/data, collection/grpc, etc

agreed. Follow-up PR? :)

**Review by dirkkul — APPROVED** [review]
